src/hypothesis/internal/conjecture/engine.py

```diff
@@ -465,24 +480,63 @@ def replace(b):
                 )
                 i += 1
 
-            i = 0
-            alternatives = None
-            while i < len(self.last_data.intervals):
-                if alternatives is None:
-                    alternatives = sorted(set(
-                        self.last_data.buffer[u:v]
-                        for u, v in self.last_data.intervals), key=len)
-                u, v = self.last_data.intervals[i]
-                for a in alternatives:
-                    buf = self.last_data.buffer
-                    if (
-                        len(a) < v - u or
-                        (len(a) == (v - u) and a < buf[u:v])
-                    ):
-                        if self.incorporate_new_buffer(buf[:u] + a + buf[v:]):
-                            alternatives = None
-                            break
-                i += 1
+            self.debug('Replacing intervals with simpler intervals')
+
+            interval_counter = -1
+            while interval_counter != self.changed:
+                interval_counter = self.changed
+                i = 0
+                alternatives = None
+                while i < len(self.last_data.intervals):
```

Comment:
There's actually an important difference, and it's that "ubiquitous use of side effects" thing again (it really is a good way to write this once you get used to the style, I promise. It's actually quite robust and the alternatives are all much more unwieldy).

last_data can change in the course of running this. When it changes, this may cause the number of intervals to change. So using hrange() here would be incorrect because you can suddenly find yourself with an index out of the range of the current set of intervals.

It is a deliberate feature that sometimes the set of intervals prior to i may have changed and we might skip them over (that's one of the reasons why running this to fixation helps) but that we still stay at the old index - it's a good heuristic for avoiding performing O(n^2) useless shrinks that would happen if we restarted the loop every time.